raiden-ts/src/transport/epics/init.ts

```diff
@@ -222,76 +222,100 @@ function fetchSortedMatrixServers$(matrixServerLookup: string, httpTimeout: numb
  */
 function setupMatrixClient$(
   server: string,
   setup: RaidenMatrixSetup | undefined,
-  { address, signer }: Pick<RaidenEpicDeps, 'address' | 'signer'>,
+  { address, signer, config$ }: Pick<RaidenEpicDeps, 'address' | 'signer' | 'config$'>,
   caps: Caps | null,
 ) {
   const serverName = getServerName(server);
   if (!serverName) throw new RaidenError(ErrorCodes.TRNS_NO_SERVERNAME, { server });
 
-  return defer(() => {
-    if (setup) {
-      // if matrixSetup was already issued before, and credentials are already in state
-      const matrix = createClient({
-        baseUrl: server,
-        userId: setup.userId,
-        accessToken: setup.accessToken,
-        deviceId: setup.deviceId,
-      });
-      return of({ matrix, server, setup });
-    } else {
-      const matrix = createClient({ baseUrl: server });
-      const userName = address.toLowerCase(),
-        userId = `@${userName}:${serverName}`;
+  return config$.pipe(
+    first(),
+    mergeMap(({ pollingInterval }) => {
+      if (setup) {
+        // if matrixSetup was already issued before, and credentials are already in state
+        const matrix = createClient({
+          baseUrl: server,
+          userId: setup.userId,
+          accessToken: setup.accessToken,
+          deviceId: setup.deviceId,
+        });
+        return of({ matrix, server, setup });
+      } else {
+        const matrix = createClient({ baseUrl: server });
+        const userName = address.toLowerCase(),
+          userId = `@${userName}:${serverName}`;
 
-      // create password as signature of serverName, then try login or register
-      return from(signer.signMessage(serverName)).pipe(
-        mergeMap((password) =>
-          from(
-            matrix.login('m.login.password', { user: userName, password, device_id: DEVICE_ID }),
-          ).pipe(catchError(() => matrix.register(userName, password))),
-        ),
-        mergeMap(({ access_token, device_id }) => {
-          // matrix.register implementation doesn't set returned credentials
-          // which would require an unnecessary additional login request if we didn't
-          // set it here, and login doesn't set deviceId, so we set all credential
-          // parameters again here after successful login or register
-          matrix.deviceId = device_id;
-          matrix._http.opts.accessToken = access_token;
-          matrix.credentials = { userId };
+        // create password as signature of serverName, then try login or register
+        return from(signer.signMessage(serverName)).pipe(
+          mergeMap((password) =>
+            retryAsync$(
+              () =>
+                matrix.login('m.login.password', {
+                  user: userName,
+                  password,
+                  device_id: DEVICE_ID,
+                }),
+              pollingInterval,
+              (err) => err?.httpStatus !== 429, // retry only if rate-limit error
+            ).pipe(
+              catchError((err) =>
```

Comment:
I know this is not part of this PR, but couldn't we check for a specific error that we are required to register first? Similar like we do above for the rate limit error.

Reply:
This doesn't happen on real life. The `register` is just a fallback for when using vanilla matrix servers, while all Raiden matrix servers have the ETH auth plugin, which only requires login and disables register. So, we first try to login with the deterministic password, and try to register if it fails, but if it fails once, then we should be good to throw and move to the next server, as the server is probably not working properly as we expect, and we want to fail early and optimize for the common case.

Reply:
Hmm okay.